### ISSUE TITLE
[Proposal] Extend connection balancer

### ISSUE THREAD

**Issue by daixiang0** [envoyproxy/envoy#20180]
*Title*: *Extend connection balancer*

*Description*:

Now we only have one balancer for connections: exact connection balancer. Since it uses Lock operations, it brings much cost, is disabled by default. It is not suitable for the case including huge connections, like acts as a gateway.

[Intel® DLB](https://networkbuilders.intel.com/solutionslibrary/queue-management-and-load-balancing-on-intel-architecture) is a PCIe device for providing load balancing features and can take responsibility for distributing system workloads for maximizing the performance. It is lock-free and can provide greater performance and lower latency than software balancers.

Now [DPDK has implemented DLB](https://doc.dpdk.org/guides/eventdevs/dlb2.html) and gets great improvement.

I think Envoy can get benefit from it, especially in the gateway cases. I will share a detailed design later and implement it as an extension. As talk with @mattklein123 offline, making the connection balancer mechanism extensible sounds ok to him.

**Comment by daixiang0** [comment]
here is the API config design:

```
  message ConnectionBalanceConfig {
    option (udpa.annotations.versioning).previous_message_type =
        "envoy.api.v2.Listener.ConnectionBalanceConfig";

    message ExactBalance {
      option (udpa.annotations.versioning).previous_message_type =
          "envoy.api.v2.Listener.ConnectionBalanceConfig.ExactBalance";
    }

    // A connection balancer implementation in extension.
    message ExtendBalance {
      string name = 1;
      string lb_policy = 2;
    }

    oneof balance_type {
      option (validate.required) = true;

      ExactBalance exact_balance = 1;

      ExtendBalance extend_balance = 2;
    }
  }

```

Config example:
```
  {
    "connection_balance_config": {
      "extend_balance": {
        "name": "dlb",
        "lb_policy": "exact"
      }
    }
  }
```

Users can inherit `ConnectionBalancer` to create a new ConnectionBalancer, belonging to `ExtendBalance`.

In `source/server/listener_impl.cc`, use `ExtendBalance` as configed.

**Comment by mattklein123** [comment]
The extension should just be a TypedExtensionConfig which will go in the oneof, but otherwise this makes sense.

**Comment by daixiang0** [comment]
fixed by #20268
